src/bootstrap/doc.rs

```diff
@@ -461,7 +461,16 @@ impl Step for Std {
         // create correct links between crates because rustdoc depends on the
         // existence of the output directories to know if it should be a local
         // or remote link.
-        let krates = ["core", "alloc", "std", "proc_macro", "test"];
+        //
+        // There's also a mild hack here where we build the first crate in this
+        // list, core, twice. This is currently necessary to make sure that
+        // cargo's cached rustc/rustdoc versions are up to date which means
+        // cargo won't delete the out_dir we create for the stampfile.
+        // Essentially any crate could go into the first slot here as it's
+        // output directory will be deleted by us (as cargo will purge the stamp
+        // file during the first slot's run), and core is relatively fast to
+        // build so works OK to fill this 'dummy' slot.
+        let krates = ["core", "core", "alloc", "std", "proc_macro", "test"];
```

Comment:
Core is unfortunately pretty chunky until https://github.com/rust-lang/rust/pull/83278 lands :/ but this seems ok as a workaround. I do think it's worth it for cargo to stop deleting the doc directory though, I'll open an issue if someone from the team doesn't comment soon.

```
$ x.py doc --stage 1 library/core
Documenting stage1 std (x86_64-unknown-linux-gnu)
Documenting core v0.0.0 (/home/joshua/rustc2/library/core)
Finished release [optimized] target(s) in 8.74s
```